dvc/repo/run.py

```diff
@@ -1,24 +1,41 @@
 from __future__ import unicode_literals
 
+import os
+from shortuuid import uuid
+
 from dvc.repo.scm_context import scm_context
 
 
 @scm_context
 def run(self, no_exec=False, **kwargs):
     from dvc.stage import Stage
 
-    with self.state:
+    with self.lock, self.state:
         stage = Stage.create(self, **kwargs)
 
-    if stage is None:
-        return None
-
-    self.check_dag(self.stages() + [stage])
-
-    with self.state:
-        if not no_exec:
-            stage.run(no_commit=kwargs.get("no_commit", False))
-
-    stage.dump()
+        if stage is None:
+            return None
+
+        self.check_dag(self.stages() + [stage])
+
+        # We are releasing locks in stage.run() to allow parallel execution. To
+        # avoid possible DAG problems, we are creating the dvc file now, so
+        # that other `dvc run`s could use it to verify DAG on their side before
+        # launching their commands. We are using random unique name for this
+        # temporary dvc file to avoid collisions in case someone is trying to
+        # create the same dvc file in another process.
+        dname = os.path.dirname(stage.path)
```

Comment:
we will have to create and extensive explanation about these temporary files and how do you cleanup them if something goes wrong. Also, I would introduce some DVC-specific convention, UUID could be not enough to differentiate them if someone decides to use UUIDs for stage names.

Reply:
@shcheklein cleanup is handled down below, user won't have to do anything, unless something gone utterly wrong that even `finally` wasn't executed.

Reply:
We don't care what naming user is using, we just care that our temp file is unique, which it is. No need to further differentiate it.

Reply:
@shcheklein In case of what type of failure? We remove that hidden temporary file if we get any errors.